feat(maps):  Added custom Leaflet TimeSlider control with React porta…

Godlisten-Ngatara opened the pull request:
## Summary
This PR introduces a TimeSlider control for DHIS2 maps as an extension to the HISPTZ library. The control is implemented as a custom Leaflet control and rendered using React portals to enable seamless integration with existing map components.

## Changes
- Created TimeSliderControl as a Leaflet control with its own container.
- Added TimeSliderControlWrapper to render the control in React via a portal.
- Registered the new control in the map’s controls configuration for easy use in MapArea.
- Ensured the implementation does not modify HISPTZ core code, maintaining backward compatibility.

## Notes
- periodSelection and orgUnitSelection integration is not yet implemented
- The feature is currently functional in isolation

## Testing
- Verified that the TimeSlider control renders correctly in Storybook.
- Confirmed that the control can be positioned using ControlPosition props.

Godlisten-Ngatara committed:
feat(maps):  Added custom Leaflet TimeSlider control with React portal rendering

nnkogift commented on packages/analytics/src/components/Map/components/MapLayer/components/TimeSlider/components/index.tsx:
```diff
@@ -0,0 +1,60 @@
+    if (playing) {
+      interval = setInterval(() => {
+        setActiveIndex((prevIndex) => (prevIndex + 1) % months.length);
+      }, 1000); // Change every 1 second
```
This should be configurable

nnkogift commented on packages/analytics/src/components/Map/DHIS2Map.stories.tsx:
```diff
@@ -35,13 +35,21 @@ Controls.args = {
+			position: "bottomright",
+			options: {
+				hidden: false,
+				hideControlContainer: false,
```
Add variable intervalTime to control the number of milliseconds a slide will be visible

nnkogift commented on packages/analytics/src/components/Map/components/MapLayer/components/TimeSlider/components/TimeSliderControlWrapper.tsx:
```diff
@@ -0,0 +1,31 @@
+export default function TimeSliderControlWrapper({
+  position,
+}: {
+  position: ControlPosition;
```
Add the intervalTime as prop

nnkogift commented on packages/analytics/src/components/Map/components/MapLayer/components/TimeSlider/components/TimeSliderControlWrapper.tsx:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <>
+      <TimeSliderLeafletControl position={position} />
+      {container && ReactDOM.createPortal(<TimeSliderLayer />, container)}
```
pass down the intervalTime  prop

nnkogift commented on packages/analytics/src/components/Map/components/MapLayer/components/TimeSlider/components/index.tsx:
```diff
@@ -0,0 +1,60 @@
+  "December",
+];
+
+export default function TimeSliderLayer() {
```
add intervalTime prop

nnkogift commented on packages/analytics/src/components/Map/DHIS2Map.stories.tsx:
```diff
@@ -35,13 +35,21 @@ Controls.args = {
+				hideControlContainer: false,
+			},
+		},
 	],
```
Add a new story with the title timelapse

nnkogift committed:
chore(analytics): integrate dynamic periods support in TimeSliderLayer

- Replaced static months array with dynamic periods fetched using `useMapPeriods` for improved flexibility.
- Updated slider logic to iterate through dynamic periods instead of a fixed list.
- Enhanced maintainability and usability by leveraging context-provided data instead of hardcoding static values.

Why: To make the TimeSliderLayer more adaptable by dynamically handling periods based on map context.

Godlisten-Ngatara committed:
feat(Map stories): Added a story for the time slider component

Godlisten-Ngatara committed:
refactor(dynamic time interval): Added a configurable dynamic interval for dynamic rendering

Godlisten-Ngatara committed:
refactor(time interval): Passed slide time interval as a prop

Godlisten-Ngatara committed:
Merge branch 'hisptz:develop' into feature/time-slider
